### ISSUE TITLE
Centering of splash screen

### ISSUE THREAD

**Issue by bollenn** [pyinstaller/pyinstaller#6271]
**Is your feature request related to a problem? Please describe.**
When multiple screens connected it seems the splash screen is centered in the complete screen instead of on the active screen.

**Describe the solution you'd like**
Center the splash screen in the active screen only

**Describe alternatives you've considered**
Configurable centering

**Additional context**
-

**Comment by bollenn** [comment]
I would say issue is in https://github.com/pyinstaller/pyinstaller/blob/ae7ed8ed044f210288b27f0c84ad670ff573822b/PyInstaller/building/splash_templates.py#L91-L95 as `winfo screenwidth .` returns the summed up width of both connected monitors.

**Comment by rokm** [comment]
Hmm, well, as far as I can tell, this is a limitation of the Tk toolkit; its `winfo` `screenwidth`/`screenheight` returns the combined desktop size. There doesn't seem to be a way to query individual monitor sizes nor determine which monitor is the primary one.

It seems [the suggested work-around for multi-screen](https://stackoverflow.com/questions/3129322/how-do-i-get-monitor-resolution-in-python/56913005#56913005) is to create a maximized/fullscreen window (possibly fully-transparent one) and read back its size. This might work for our splash screen as well...

**Comment by bwiecek1** [comment]
The solution to this problem is system specific. For Windows I have modified the pyinstaller (based on v6.19.0) to use twapi package for centering the splash on the appropriate monitor. It defaults to standard behavior if no twapi package is available. It would be as simple as adding the twapi and dependent registry packages to the tcl_data folder, then modifying the "splash_templates.py" file to require twapi package and adding tcl code to center the splash image, but I found out that the "auto_path" variable is mishandled in the bootloader. Tcl treats the "auto_path" as a list object and the bootloader assumes a simple string variable resulting in a mangled tcl_data folder path (tcl_modules_dir). I had to modify the bootloader source files and recompile it for this to work.

**Comment by rokm** [comment]
>  but I found out that the "auto_path" variable is mishandled in the bootloader. Tcl treats the "auto_path" as a list object and the bootloader assumes a simple string variable resulting in a mangled tcl_data folder path (tcl_modules_dir). I had to modify the bootloader source files and recompile it for this to work.

Well-spotted! We'll fix that.

As for the Windows-specific solution, I suppose we could have bootloader perform the equivalent win32 API calls to query resolution of the primary monitor (presumably `GetSystemMetrics()` with `SM_CXFULLSCREEN` and `SM_CYFULLSCREEN`) and store the values in global Tcl variables for use in the splash screen script. That would be a lot cleaner and easier than trying to bundle `twapi` and its dependencies (and ensuring they are properly extracted in `onefile` mode).

**Comment by bwiecek1** [comment]
Yes, bootloader win32 API calls is the way to go, I just did not want to touch the bootloader initially at all but have a quick and dirty manual solution with simply copying the appropriate packages and modifying tcl script. As always, it did not work out that way. 
For my solution here are the code snippets:

The relevant portion of the tcl script using twapi package

splash_canvas_setup = r"""
package require Tk

set display_width [winfo screenwidth .]
set display_height [winfo screenheight .]

set twapi_code [ catch {package require twapi} result ]
if {$twapi_code == 0} {
    # If twapi is available
    set pos [twapi::get_mouse_location]
    set px [lindex $pos 0]
    set py [lindex $pos 1]
    set monitor [twapi::get_display_monitor_from_point $px $py -default nearest]
    set monitor_info [twapi::get_display_monitor_info $monitor]
    set monitor_extent [dict get $monitor_info -extent]
    set monitor_posx [lindex $monitor_extent 0]
    set monitor_posy [lindex $monitor_extent 1]
    set display_width [expr {[lindex $monitor_extent 2] - $monitor_posx}]
    set display_height [expr {[lindex $monitor_extent 3] - $monitor_posy}]
} else {
    # If twapi is not available, we fallback to the primary monitor
    set monitor_posx 0
    set monitor_posy 0
}

set image_width [image width splash_image]
set image_height [image height splash_image]

set x_position [expr {int(0.5*($display_width - $image_width)+$monitor_posx)}]
set y_position [expr {int(0.5*($display_height - $image_height)+$monitor_posy)}]
(...)

and here is the C code equivalent to obtain the same information:

int monitor_posx = 0;
int monitor_posy = 0;
int monitor_width = -1;
int monitor_height = -1;
POINT mouse_point;

if (GetCursorPos(&mouse_point))
{
	HMONITOR hm = MonitorFromPoint(mouse_point, MONITOR_DEFAULTTONEAREST);
	if (hm != NULL) /* in most cases it should always be non-NULL */
	{
		MONITORINFO minfo;
		if (GetMonitorInfoA(hm, &minfo))
		{
			monitor_posx = minfo.rcMonitor.left;
			monitor_posy = minfo.rcMonitor.top;
			monitor_width = minfo.rcMonitor.right - monitor_posx;
			monitor_height  = minfo.rcMonitor.bottom - monitor_posy;
		}
	}
}

I set the width and height to negative numbers to indicate failure in obtaining monitor information.
One could use rcWork rectangle rather than rcMonitor to calculate position and centering in the work area rather than the whole  monitor area.

**Comment by bwiecek1** [comment]
I missed the fact that code above was not formatted. Here's the C code that I have for creating _monitor_geometry variable on WIN32 systems

~~~
#if defined(_WIN32)   
     /* 
      *Variables for monitor information to place splash image
      * on an appropriate monitor so that the splash screen appears
      * centered on the screen where the application is running. 
      */
    int monitor_posx = 0;
    int monitor_posy = 0;
    int monitor_width = -1;
    int monitor_height = -1;
    POINT mouse_point;
    /* 
     * Determine the monitor on which the splash screen should appear
     * by querying the current mouse position and the monitor
     * that contains that point.
     */
    if (GetCursorPos(&mouse_point)) {
        HMONITOR monitor = MonitorFromPoint(mouse_point, MONITOR_DEFAULTTONEAREST);
        if (monitor) {
            MONITORINFO mi;
            mi.cbSize = sizeof(mi);
            if (GetMonitorInfo(monitor, &mi)) {
                monitor_posx = mi.rcMonitor.left;
                monitor_posy = mi.rcMonitor.top;
                monitor_width = mi.rcMonitor.right - mi.rcMonitor.left;
                monitor_height = mi.rcMonitor.bottom - mi.rcMonitor.top;
            }
        }
    }
    /* If we were able to determine the monitor bounds, we can
     * pass them to Tcl/Tk so that the splash screen can be
     * positioned appropriately. */
    if (monitor_width > 0 && monitor_height > 0) {
        char geom[64];
        snprintf(geom, sizeof(geom), "%dx%d%+d%+d",
                 monitor_width, monitor_height,
                 monitor_posx, monitor_posy);
        dylib_tcltk->Tcl_SetVar2(splash->interp, "_monitor_geometry", NULL, geom, TCL_GLOBAL_ONLY);
    }
#endif
~~~

and the associated tcl script snippet:

~~~
splash_canvas_setup = r"""
package require Tk

set display_width [winfo screenwidth .]
set display_height [winfo screenheight .]
set display_posx 0
set display_posy 0

if {[info exists _monitor_geometry]} {
    # If the bootloader passed monitor geometry, use it
    set code [ catch {scan $_monitor_geometry "%dx%d%d%d" monitor_width monitor_height monitor_posx monitor_posy} result ]
    # If scan succeeded, code will be 0 and monitor geometry
    # values will be set in monitor_width/height/posx/posy
    if {$code == 0} {
        set display_posx $monitor_posx
        set display_posy $monitor_posy
        set display_width $monitor_width
        set display_height $monitor_height
    }
}

set image_width [image width splash_image]
set image_height [image height splash_image]

set x_position [expr {int(0.5*($display_width - $image_width) + $display_posx)}]
set y_position [expr {int(0.5*($display_height - $image_height) + $display_posy)}]

# Toplevel frame in which all widgets should be positioned
frame .root
...
~~~

**Comment by rokm** [comment]
Initial support for alternative/explicit splash screen centering modes was added in #9425.
